File: java/src/edu/uf/intracellularState/FMacrophageBooleanNetwork.java
package edu.uf.intracellularState;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import edu.uf.interactable.Hepcidin;
import edu.uf.interactable.IFN_I;
import edu.uf.interactable.IFN_II;
import edu.uf.interactable.IL1;
import edu.uf.interactable.IL10;
import edu.uf.interactable.IL4;
import edu.uf.interactable.Molecule;
import edu.uf.interactable.PtSRBinder;
import edu.uf.interactable.TGFb;
import edu.uf.interactable.TLRBinder;
import edu.uf.interactable.TNFa;
import edu.uf.interactable.Afumigatus.Afumigatus;
import edu.uf.interactable.invitro.BGlucan;
//import edu.uf.interactable.covid.SAMP;
import edu.uf.interactable.invitro.GM_CSF;
import edu.uf.utils.Constants;

public abstract class FMacrophageBooleanNetwork extends IntracellularModel{
	
	public static final int size = 29;
	//public static final int NUM_RECEPTORS = 12;
	
	public static final int IFNGR = 0;
	public static final int IFNB = 1;
	public static final int CSF2Ra = 2;
	public static final int IL1R = 3;
	public static final int IL1B = 4;
	public static final int TLR4 = 5;
	public static final int FCGR = 6;
	public static final int IL4Ra = 7;
	public static final int IL10R = 8;
	public static final int IL10_out = 9;
	public static final int STAT1 = 10;
	public static final int SOCS1 = 11;
	public static final int STAT3 = 12;
	public static final int STAT5 = 13;
	public static final int IRF4 = 14;
	public static final int NFkB = 15;
	public static final int PPARG = 16;
	public static final int KLF4 = 17;
	public static final int STAT6 = 18;
	public static final int JMJD3 = 19;
	public static final int IRF3 = 20;
	public static final int ERK = 21;
	public static final int IL12_out = 22;
	public static final int TNFR = 23;
	public static final int Dectin = 24;
	public static final int ALK5 = 25;
	public static final int SMAD2 = 26;
	public static final int PtSR = 27;
	public static final int FPN = 28;
	
	
	/*public static final int M1  = Phenotype.createPhenotype();
	public static final int M2A = Phenotype.createPhenotype();
	public static final int M2B = Phenotype.createPhenotype();
	public static final int M2C = Phenotype.createPhenotype();*/
	
	//public static final int FNP = 28; //IN THE OUTER CLASS!
	
	{
		this.booleanNetwork = new int[size];
	}
	
	private static final int N = 4;
	
	@Override
	public void processBooleanNetwork(int... args) {
		
		int k = 0;
		List<Integer> array = new ArrayList<>(size);
		for(int i = 0; i < size; i++)
			array.add(i);
		while(true) {
			if(k++ > Constants.MAX_BN_ITERATIONS)break;
			Collections.shuffle(array, new Random());
			for(int i : array) {
				switch(i) {
					case 0:
						this.booleanNetwork[IFNGR] = max(this.booleanNetwork[IFNB], getInput(Molecule.searchMolecule(IFN_II.NAME)));
						break;
					case 1:
						this.booleanNetwork[CSF2Ra] = getInput(Molecule.searchMolecule("GM_CSF"));
						break;
					case 2:
						this.booleanNetwork[IL1R] = max(this.booleanNetwork[IL1B], getInput(Molecule.searchMolecule(IL1.NAME)));
						break;
					case 3:
						this.booleanNetwork[TLR4] = min(getInput(TLRBinder.getBinder()), not(this.booleanNetwork[FCGR], N));
						break;
					case 4:
						this.booleanNetwork[FCGR] = max(
								min(getInput(Molecule.searchMolecule("IC")), getInput(TLRBinder.getBinder())),
								min(getInput(Molecule.searchMolecule("IC")), getInput(Molecule.searchMolecule(IL1.NAME)))
						);
						break;
					case 5:
						this.booleanNetwork[IL4Ra] = getInput(Molecule.searchMolecule(IL4.NAME));
						break;
					case 6:
						this.booleanNetwork[IL10R] = max(getInput(Molecule.searchMolecule(IL10.NAME)), this.booleanNetwork[IL10_out]);
						break;
					case 7:
						this.booleanNetwork[STAT1] = min(this.booleanNetwork[IFNGR], not(max(this.booleanNetwork[SOCS1], this.booleanNetwork[STAT3]), N));
						break;
					case 8:
						this.booleanNetwork[STAT5] = min(this.booleanNetwork[CSF2Ra], not(max(this.booleanNetwork[STAT3], this.booleanNetwork[IRF4]), N));
						break;
					case 9:
						this.booleanNetwork[NFkB] = min(max(new int[] {
								this.booleanNetwork[IL1R], 
								this.booleanNetwork[TLR4], 
								this.booleanNetwork[Dectin], 
								this.booleanNetwork[TNFR]
						}), not(max(new int[] {
								this.booleanNetwork[STAT3], 
								this.booleanNetwork[FCGR], 
								this.booleanNetwork[PPARG], 
								this.booleanNetwork[KLF4]}), N));
						break;
					case 10:
						this.booleanNetwork[PPARG] = this.booleanNetwork[IL4Ra];
						break;
					case 11:
						this.booleanNetwork[STAT6] = this.booleanNetwork[IL4Ra];
						break;
					case 12:
						this.booleanNetwork[JMJD3] = this.booleanNetwork[IL4Ra];
						break;
					case 13:
						this.booleanNetwork[STAT3] = min(max(this.booleanNetwork[IL10R], this.booleanNetwork[SMAD2]), not(
								max(this.booleanNetwork[FCGR], this.booleanNetwork[PPARG]), N));
						break;
					case 14:
						this.booleanNetwork[IRF3] = this.booleanNetwork[TLR4];
						break;
					case 15:
						this.booleanNetwork[ERK] = this.booleanNetwork[FCGR];// | this.booleanNetwork[Dectin];
						break;
					case 16:
						this.booleanNetwork[KLF4] = this.booleanNetwork[STAT6];
						break;
					case 17:
						this.booleanNetwork[IL1B] = this.booleanNetwork[NFkB];
						break;
					case 18:
						this.booleanNetwork[IFNB] = this.booleanNetwork[IRF3];
						break;
					case 19:
						this.booleanNetwork[IL12_out] = max(new int[] {this.booleanNetwork[STAT1], this.booleanNetwork[STAT5], this.booleanNetwork[NFkB]});
						break;
					case 20: 
						this.booleanNetwork[IL10_out] = max(new int[] {
								this.booleanNetwork[PPARG], 
								this.booleanNetwork[STAT6], 
								this.booleanNetwork[JMJD3], 
								this.booleanNetwork[STAT3], 
								this.booleanNetwork[ERK], 
								this.booleanNetwork[PtSR]
						});
						break;
					case 21:
						this.booleanNetwork[TNFR] = getInput(Molecule.searchMolecule(TNFa.NAME));
						break;
					case 22:
						this.booleanNetwork[Dectin] = max(getInput(Afumigatus.DEF_OBJ), getInput(Molecule.searchMolecule(BGlucan.NAME)));
						break;
					case 23:
						this.booleanNetwork[SOCS1] = this.booleanNetwork[STAT6];
						break;
					case 24:
						this.booleanNetwork[IRF4] = this.booleanNetwork[JMJD3];
						break;
					case 25:
						this.booleanNetwork[ALK5] = getInput(Molecule.searchMolecule(TGFb.NAME));
						break;
					case 26:
						this.booleanNetwork[SMAD2] = this.booleanNetwork[ALK5];
						break;
					case 27:
						this.booleanNetwork[PtSR] = getInput(PtSRBinder.getBinder());
						break;
					case 28:
						//this.booleanNetwork[FPN] = (-getInput(Hepcidin.getMolecule()) + N);
						break;
					default:
						System.err.println("No such interaction " + i + "!");
						break;
				}
			}
		}
		
		this.inputs.clear();
		this.clearPhenotype();
		this.computePhenotype();
		
		
		/*if(this.booleanNetwork[NFkB] > 0 || this.booleanNetwork[STAT1] > 0 || this.booleanNetwork[STAT5] > 0)
			this.getPhenotype().put(M1, this.or(new int[] {this.booleanNetwork[NFkB], this.booleanNetwork[STAT1], this.booleanNetwork[STAT5]}));
		if(this.booleanNetwork[STAT6] > 0)
			this.getPhenotype().put(M2A, this.booleanNetwork[STAT6]);
		if(this.booleanNetwork[ERK] > 0)
			this.getPhenotype().put(M2B, this.booleanNetwork[ERK]);
		if(this.booleanNetwork[STAT3] > 0)
			this.getPhenotype().put(M2C, this.booleanNetwork[STAT3]);
		*/
		//else
		//	Macrophage.this.addPhenotype(Phenotypes.RESTING);
			
		
	
		
	}

}

File: cpp/src/edu/uf/intracellularState/FMacrophageBooleanNetwork.cpp
/*
 * FMacrophageBooleanNetwork.cpp
 *
 *  Created on: Jul 22, 2024
 *      Author: henriquedeassis
 */

#include "FMacrophageBooleanNetwork.h"
#include "../interactable/IL10.h"
#include "../interactable/TGFb.h"
#include "../interactable/TNFa.h"
#include "../interactable/afumigatus/Afumigatus.h"
#include "../utils/Constants.h"
#include <vector>
#include <algorithm>
#include <random>
#include <iostream>

namespace edu {
namespace uf {
namespace intracellularState {

using namespace edu::uf::utils;
using namespace afumigatus;


void FMacrophageBooleanNetwork::processBooleanNetwork() {
    int k = 0;
    vector<int> array(size);
    iota(array.begin(), array.end(), 0);

    while (true) {
        if (k++ > utils::constexprants::MAX_BN_ITERATIONS) break;
        shuffle(array.begin(), array.end(), Rand::getRand().getGenerator());
        for (int i : array) {
            switch (i) {
                case IFNGR:
                    //booleanNetwork[IFNGR] = booleanNetwork[IFNB];
                    break;
                case CSF2Ra:
                    //booleanNetwork[CSF2Ra] = getInput(GM_CSF::getMolecule());
                    break;
                case IL1R:
                    //booleanNetwork[IL1R] = max(booleanNetwork[IL1B], getInput(IL1::getMolecule()));
                    break;
                case TLR4:
                    //booleanNetwork[TLR4] = min(getInput(TLRBinder::getBinder()), gnot(booleanNetwork[FCGR], N));
                	booleanNetwork[TLR4] = min(getInput(Afumigatus::DEF_OBJ), gnot(booleanNetwork[FCGR], N));
                    break;
                case FCGR:
                    booleanNetwork[FCGR] = 0;
                    break;
                case IL4Ra:
                    //booleanNetwork[IL4Ra] = getInput(IL4::getMolecule());
                    break;
                case IL10R:
                    booleanNetwork[IL10R] = max(getInput(IL10::getMolecule()), booleanNetwork[IL10_out]);
                    break;
                case STAT1:
                    booleanNetwork[STAT1] = min(booleanNetwork[IFNGR], gnot(max(booleanNetwork[SOCS1], booleanNetwork[STAT3]), N));
                    break;
                case STAT5:
                    booleanNetwork[STAT5] = min(booleanNetwork[CSF2Ra], gnot(max(booleanNetwork[STAT3], booleanNetwork[IRF4]), N));
                    break;
                case NFkB:{
                    int nfkbInputs[] = { booleanNetwork[IL1R], booleanNetwork[TLR4], booleanNetwork[Dectin], booleanNetwork[TNFR] };
                    int nfkbInhibitors[] = { booleanNetwork[STAT3], booleanNetwork[FCGR], booleanNetwork[PPARG], booleanNetwork[KLF4] };
                    booleanNetwork[NFkB] = min(amax(nfkbInputs, 4), gnot(amax(nfkbInhibitors, 4), N));
                    break;
                }
                case PPARG:
                    booleanNetwork[PPARG] = booleanNetwork[IL4Ra];
                    break;
                case STAT6:
                    booleanNetwork[STAT6] = booleanNetwork[IL4Ra];
                    break;
                case JMJD3:
                    booleanNetwork[JMJD3] = booleanNetwork[IL4Ra];
                    break;
                case STAT3:
                    booleanNetwork[STAT3] = min(max(booleanNetwork[IL10R], booleanNetwork[SMAD2]), gnot(max(booleanNetwork[FCGR], booleanNetwork[PPARG]), N));
                    break;
                case IRF3:
                    booleanNetwork[IRF3] = booleanNetwork[TLR4];
                    break;
                case ERK:
                    booleanNetwork[ERK] = booleanNetwork[FCGR];
                    break;
                case KLF4:
                    booleanNetwork[KLF4] = booleanNetwork[STAT6];
                    break;
                case IL1B:
                    booleanNetwork[IL1B] = booleanNetwork[NFkB];
                    break;
                case IFNB:
                    booleanNetwork[IFNB] = booleanNetwork[IRF3];
                    break;
                case IL12_out:{
                    int il12_outInputs[] = { booleanNetwork[STAT1], booleanNetwork[STAT5], booleanNetwork[NFkB] };
                    booleanNetwork[IL12_out] = amax(il12_outInputs, 3);
                    break;
                }
                case IL10_out:{
                    int il10_outInputs[] = { booleanNetwork[PPARG], booleanNetwork[STAT6], booleanNetwork[JMJD3], booleanNetwork[STAT3], booleanNetwork[ERK], booleanNetwork[PtSR] };
                    booleanNetwork[IL10_out] = amax(il10_outInputs, 6);
                    break;
                }
                case TNFR:
                    booleanNetwork[TNFR] = getInput(TNFa::getMolecule());
                    break;
                case Dectin:
                    booleanNetwork[Dectin] = getInput(Afumigatus::DEF_OBJ);
                    break;
                case SOCS1:
                    booleanNetwork[SOCS1] = booleanNetwork[STAT6];
                    break;
                case IRF4:
                    booleanNetwork[IRF4] = booleanNetwork[JMJD3];
                    break;
                case ALK5:
                    booleanNetwork[ALK5] = getInput(TGFb::getMolecule());
                    break;
                case SMAD2:
                    booleanNetwork[SMAD2] = booleanNetwork[ALK5];
                    break;
                case PtSR:
                    // booleanNetwork[PtSR] = getInput(SAMP::getMolecule());
                    break;
                case FPN:
                    //booleanNetwork[FPN] = gnot(getInput(Hepcidin::getMolecule()), N);
                    break;
                default:
                    cerr << "No such interaction " << i << "!" << endl;
                    break;
            }
        }
    }

    inputs.clear();
    clearPhenotype();
    computePhenotype();
}

} // namespace intracellularState
} // namespace uf
} // namespace edu
